docs/source/federation/gateway.mdx

```diff
@@ -0,0 +1,227 @@
+```
+
+To implement this flow, we can use the `didReceiveResponse` callback of the `RemoteGraphQLDataSource` class to inspect each implementing service's result as it comes in. We can add the `Server-Id` to the shared `context` in this callback, then pull the full list from the `context` when sending the final response to the client.
+
+<ExpansionPanel title="Expand example">
+
+```javascript
+const { ApolloServer } = require('apollo-server');
+const { ApolloGateway, RemoteGraphQLDataSource } = require('@apollo/gateway');
+
+class DataSourceWithServerId extends RemoteGraphQLDataSource {
+  // highlight-start
+  async didReceiveResponse({ response, request, context }) {
+    // Parse the Server-Id header and add it to the array on context
+    const serverId = response.http.headers.get('Server-Id');
+    if (serverId) {
+      context.serverIds.push(serverId);
+    }
+
+    // Return the response, even when unchanged.
+    return response;
+  }
+  // highlight-end
+}
+
+const gateway = new ApolloGateway({
+  serviceList: [
+    { name: 'products', url: 'http://localhost:4001' }
+    // other services
+  ],
+  // highlight-start
+  buildService({ url }) {
+    return new DataSourceWithServerId({ url });
+  }
+  // highlight-end
+});
+
+const server = new ApolloServer({
+  gateway,
+  subscriptions: false, // Must be disabled with the gateway; see above.
+  context() {
+    return { serverIds: [] };
+  },
+  plugins: [
+    // highlight-start
+    {
+      requestDidStart() {
+        return {
+          willSendResponse({ context, response }) {
+            // Append our final result to the outgoing response headers
+            response.http.headers.append(
+              'Server-Id',
+              context.serverIds.join(',')
+            );
+          }
+        };
+      }
+    }
+    // highlight-end
+  ]
+});
+
+server.listen().then(({ url }) => {
+  console.log(`🚀 Server ready at ${url}`);
+});
+```
+
+</ExpansionPanel>
+
+To learn more about `buildService` and `RemoteGraphQLDataSource`, see the [API docs](/api/apollo-gateway/).
+
+## Implementing custom directives
+
+The `@apollo/gateway` library provides limited support for custom [directives](/schema/directives/) that are implemented by your implementing services. Directives must meet all of the following criteria to be compatible with the gateway:
+
+* Custom directives can only be used in **executable locations**, as defined in the [GraphQL specification](https://graphql.github.io/graphql-spec/June2018/#ExecutableDirectiveLocation).
+* Your custom directives must be defined by *every* implementing service in your data graph, even if not all of your services use them.
+* Directive definitions must be identical across all of your service schemas. Directive definitions are identical if their names, locations, arguments, and argument types are all identical.
+
+>Apollo Server does not currently support executable directives, however they are supported by the gateway.
```

Comment:
This seems like it could be more prominent in this section — immediately below the heading, and even possibly just suggest that the intended target for this implementation suggestion is either users who have explicitly used `graphql-tools` tooling to augment a schema with directives or have built a schema in another way other than using Apollo Server's normal `typeDefs` behavior, or those who are using a completely different downstream server written in another language with Apollo Gateway.